cmd/cli/update.go

```diff
@@ -0,0 +1,178 @@
+guidelines as a valid Kubernetes label value. Must be 63 characters or
+less and must be empty or begin and end with an alphanumeric character
+([a-z0-9A-Z]) with dashes (-), underscores (_), dots (.), and
+alphanumerics between.
+
+Example:
+  $ osm update --mesh-name "hello-osm"
+
+The mesh name is used in various ways like for naming Kubernetes resources as
+well as for adding a Kubernetes Namespace to the list of Namespaces a control
+plane should watch for sidecar injection of Envoy proxies.
+`
+
+type updateCmd struct {
+	out       io.Writer
+	clientSet kubernetes.Interface
+
+	chartPath                     string
+	enablePermissiveTrafficPolicy bool
+	enableEgress                  bool
+	meshCIDRRanges                []string
+}
+
+func newUpdateCmd(config *helm.Configuration, out io.Writer) *cobra.Command {
+	upd := &updateCmd{
+		out: out,
+	}
+
+	cmd := &cobra.Command{
+		Use:   "update",
+		Short: "update osm control plane configuration",
+		Long:  updateDesc,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			kubeconfig, err := settings.RESTClientGetter().ToRESTConfig()
+			if err != nil {
+				return errors.Errorf("Error fetching kubeconfig")
+			}
+
+			clientset, err := kubernetes.NewForConfig(kubeconfig)
+			if err != nil {
+				return errors.Errorf("Could not access Kubernetes cluster. Check kubeconfig")
+			}
+			upd.clientSet = clientset
+			return upd.run(config)
+		},
+	}
+
+	f := cmd.Flags()
+
+	f.StringVar(&upd.chartPath, "osm-chart-path", "", "path to osm chart to override default chart")
+	f.BoolVar(&upd.enablePermissiveTrafficPolicy, "enable-permissive-traffic-policy", false, "Enable permissive traffic policy mode")
+	f.BoolVar(&upd.enableEgress, "enable-egress", false, "Enable egress in the mesh")
+	f.StringSliceVar(&upd.meshCIDRRanges, "mesh-cidr", []string{}, "mesh CIDR range, accepts multiple CIDRs, required if enable-egress option is true")
+
+	return cmd
+}
+
+func (u *updateCmd) run(config *helm.Configuration) error {
+	var chartRequested *chart.Chart
+	var err error
+	if u.chartPath != "" {
+		chartRequested, err = loader.Load(u.chartPath)
+	} else {
+		chartRequested, err = cli.LoadChart(chartTGZSource)
+	}
+	if err != nil {
+		return err
+	}
+
+	// Validate CIDR ranges if egress is enabled
+	if u.enableEgress {
+		if err := validateCIDRs(u.meshCIDRRanges); err != nil {
+			return errors.Errorf("Invalid mesh-cidr-ranges: %q, error: %v. Valid mesh CIDR ranges must be specified with egress enabled.", u.meshCIDRRanges, err)
+		}
+	}
+
+	values, err := u.resolveValues()
+	if err != nil {
+		return err
+	}
+
+	// TODO (nitishm): Should we check if the controlplane deployment exists ?
+
+	upgradeClient := helm.NewUpgrade(config)
+	upgradeClient.Namespace = settings.Namespace()
+	if _, err = upgradeClient.Run(chartRequested.Name(), chartRequested, values); err != nil {
+		return err
+	}
+
+	fmt.Fprintf(u.out, "OSM successfully updated in namespace [%s]\n", settings.Namespace())
```

Comment:
@michelleN If we need to also print the Mesh Name we will need to do a lookup.  WDYT?

Reply:
> Thanks. Left some comments inline.
> On another note: The issue defined was for `osm update MESH_NAME` but now that I am reading through it, I'm thinking we should nest the update command under the `osm mesh` command so: `osm mesh update MESH_NAME` what do you think?

IMO I agree. All mesh-related commands would best be under the `mesh`. This would include `install`, `delete` and `update`. Is that what you are thinking as well?

Reply:
I think install can remain top level as it is a command that will be used frequently. All other mesh related commands should be nested under `osm mesh`

Reply:
Sounds good. Do you want me to do it in this PR or a follow up PR?

Reply:
In this PR please.